File: Arrays/442. Find All Duplicates in an Array/442. Find All Duplicates in an Array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findDuplicates = function (nums) {
  ans = [];
  for (let i = 0; i < nums.length; i++) {
    const idx = Math.abs(nums[i]) - 1;
    if (nums[idx] < 0) ans.push(idx + 1);
    else nums[idx] *= -1;
  }
  return ans;
};

// Time: O(n)
// Space: O(1)

File: Arrays/442. Find All Duplicates in an Array/442. Find All Duplicates in an Array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        ans = []
        for i, n in enumerate(nums):
            idx = abs(n) - 1
            if nums[idx] < 0:
                ans.append(idx + 1)
            else:
                nums[idx] *= -1
        
        return ans

# Time: O(n)
# Space: O(1)